### PR TITLE
Added and Configured dev tools

### LINKED ISSUE #3: Set up linting tools and code formatter
Set up the below tools:
- [x] ES Lint
- [x] Prettier
- [x] Commit Lint
- [x] Husky

### PR DESCRIPTION
**Pull request by Pranav-Patani**
This PR resolves #3 
Dev tools include:
- prettier
- commitlint
- eslint
- husky

### CODE REVIEW

**Comment by khusharth** [comment]
Issues will only be autoclosed if using one of the following keywords:
close
closes
closed
fix
fixes
fixed
resolve
resolves
resolved

https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

File: package.json [hunk 1]
```diff
@@ -4,13 +4,21 @@
...
   "scripts": {
     "start": "parcel index.html",
-    "build": "parcel build index.html --dist-dir ./dist"
+    "build": "parcel build index.html --dist-dir ./dist",
+    "prepare": "husky"
   },
   "author": "Pranav Patani",
   "license": "ISC",
   "devDependencies": {
+    "@commitlint/cli": "^19.3.0",
```
Comment (khusharth) [inline package.json:13]: 1. is commit lint running on every commit?
1. All the files that are present in the staging area when committing - are we running eslint and prettier on them to check for errors? - Commit should fail in case of any formatting or eslint error too. Can check this package - https://github.com/lint-staged/lint-staged

File: package.json [hunk 2]
```diff
@@ -17,8 +17,19 @@
...
     "globals": "^15.8.0",
     "husky": "^9.0.11",
+    "lint-staged": "^15.2.7",
     "parcel": "2.12",
     "prettier": "^3.3.2",
     "sass": "^1.77.6"
+  },
+  "lint-staged": {
+    "*.js": [
+      "prettier --write",
```
Comment (khusharth) [inline package.json:27]: why using --write here? Do you want it to automatically format on commit?

File: package.json [hunk 3]
```diff
@@ -17,8 +17,19 @@
...
+      "eslint --fix",
+      "eslint"
```
Comment (khusharth) [inline package.json:29]: why using both eslint and eslint --fix?

File: package.json [hunk 4]
```diff
@@ -17,8 +17,19 @@
     "eslint": "^9.6.0",
     "globals": "^15.8.0",
     "husky": "^9.0.11",
+    "lint-staged": "^15.2.7",
     "parcel": "2.12",
     "prettier": "^3.3.2",
     "sass": "^1.77.6"
+  },
+  "lint-staged": {
```
Comment (khusharth) [inline package.json:25]: what about css? no need to use prettier on CSS files?

Also checkout https://stylelint.io/ for linting css

File: eslint.config.mjs [hunk 5]
```diff
@@ -1,27 +1,10 @@
...
-  extends: 'eslint:recommended',
-  parserOptions: {
-    ecmaVersion: 6
-  },
-  env: {
-    node: true,
-    browser: true,
-    es6: true
-  }
-};
```
Comment (khusharth) [inline eslint.config.mjs]: why was this removed?

File: package.json [hunk 6]
```diff
@@ -17,8 +17,19 @@
     "eslint": "^9.6.0",
     "globals": "^15.8.0",
     "husky": "^9.0.11",
+    "lint-staged": "^15.2.7",
```
Comment (khusharth) [inline package.json:20]: why are these packages as a dev dependency and not a normal dependency?

File: .husky/pre-commit [hunk 7]
```diff
@@ -1 +1,4 @@
+#!/usr/bin/env.sh
+. "$(dirname -- "$0")/_/husky.sh"
 
+npx lint-staged
```
Comment (khusharth) [inline .husky/pre-commit:4]: Prefer to use npm vs npx.

Figure out why we should do that and explain first

**Comment by khusharth** [comment]
After solving a comment, you should also reply to the comment so that the reviewer knows which comment got solved

File: package.json (hunk 2 above)
Comment (Pranav-Patani) [reply package.json:31]: Yes, as it is just formatting the code, the logic will not get affected so it is more efficient to format the code automatically rather than doing it manually

File: package.json (hunk 3 above)
Comment (Pranav-Patani) [reply package.json]: Fixed, I thought eslint will list the changes required and eslint --fix will fix them, but eslint --fix lists them and changes them as well

File: package.json (hunk 4 above)
Comment (Pranav-Patani) [reply package.json:29]: Fixed, added .sass extension instead, as css file will be auto-generated and compressed

File: eslint.config.mjs (hunk 5 above)
Comment (Pranav-Patani) [reply eslint.config.mjs]: There is no need to manually write these configurations as their default values are the latest versions, and also this format was incorrect with the latest version

File: package.json (hunk 6 above)
Comment (Pranav-Patani) [reply package.json:21]: A normal dependency should contain only the packages that are required during the production phase, as all of these packages are only required during the development phase, adding them as a dev-dependency will reduce some load during the production phase.

File: .husky/pre-commit (hunk 7 above)
Comment (Pranav-Patani) [reply .husky/pre-commit]: Changed to npm.

NPM is a package manager, hence it will locally install the packages before running them so their versions can be managed on a local level.

NPX is a package executor, hence it directly fetches the latest version of the package from the npm registry and executes them. It is good when we want to execute something just once.

In my case, it is better to install lint-staged locally so the version will stay consistent and any bugs from the latest versions will be avoided as well.

Thanks for pointing it out.

File: .husky/commit-msg [hunk 8]
```diff
@@ -0,0 +1 @@
+npx --no -- commitlint --edit $1
```
Comment (khusharth) [inline .husky/commit-msg:1]: npx?

**Comment by khusharth** [comment]
Please check if the target branch is correct. Do you want to merge these in file_structure branch?

**Comment by khusharth** [comment]
Also in the diff I can see files related to sass getting deleted, but they were never added before. Why are they showing up here then?

Screenshot 2024-07-10 at 12 11 28 PM

File: package.json (hunk 6 above)
Comment (khusharth) [reply package.json:21]: what do you mean by **_production phase_** and **_load_** during production phase

**Comment by khusharth** [comment]
Consider squashing commits like testing one together - https://www.git-tower.com/learn/git/faq/git-squash

**Comment by Pranav-Patani** [comment]
> Please check if the target branch is correct. Do you want to merge these in file_structure branch?

Yes, while creating the dev_tools branch, I stacked it under file_structure branch, so now it will be merged in that order.

**Comment by Pranav-Patani** [comment]
> Also in the diff I can see files related to sass getting deleted, but they were never added before. Why are they showing up here then?
> 
> Screenshot 2024-07-10 at 12 11 28 PM

These were added in its parent, file_structure branch, so when I created the child branch these changes were carried.
After adding stylelint I got to know it does not allow empty files to be committed so had to delete them.

Changes in dev_tools can overwrite file_structure branch as it has both the structure and tools set up, then the sass files will be added with the next branch with basic css applied.

**Comment by khusharth** [comment]
> > Please check if the target branch is correct. Do you want to merge these in file_structure branch?
> 
> Yes, while creating the dev_tools branch, I stacked it under file_structure branch, so now it will be merged in that order.

But `file_structure` is already merged in `main` right?

**Comment by khusharth** [comment]
> > Also in the diff I can see files related to sass getting deleted, but they were never added before. Why are they showing up here then?
> > Screenshot 2024-07-10 at 12 11 28 PM
> 
> These were added in its parent, file_structure branch, so when I created the child branch these changes were carried. After adding stylelint I got to know it does not allow empty files to be committed so had to delete them.
> 
> Changes in dev_tools can overwrite file_structure branch as it has both the structure and tools set up, then the sass files will be added with the next branch with basic css applied.

But if these changes were already part of a branch which is merged in main then you should not need to delete them as they will not show in diff? 🤔

File: package.json (hunk 1 above)
Comment (Pranav-Patani) [reply package.json:15]: Fixed, [commit reference](https://github.com/Pranav-Patani/BookWise/pull/4/commits/7033b4cda6823c2c4d9c68bd53d1c7b2fba95d13)

File: package.json (hunk 6 above)
Comment (Pranav-Patani) [reply package.json:22]: The production phase is the phase during which the website is live; hosted on a server where it is being accessed by the users in the real time.

By load during this phase I am referring to the download time it will take and also the resources it will consume, like memory, cpu etc due to the packages downloaded.

That's why keeping only crucial dependencies for this phase is important as it will affect the speed of the website.

File: .husky/commit-msg (hunk 8 above)
Comment (Pranav-Patani) [reply .husky/commit-msg]: fixed https://github.com/Pranav-Patani/BookWise/pull/4/commits/806bf038899c9465daa2da2c48cb413fe17fd362

File: package.json (hunk 1 above)
Comment (Pranav-Patani) [reply package.json:15]: Fixed https://github.com/Pranav-Patani/BookWise/pull/4/commits/40f1e2f1f54a7d2fb534decc769a0bc7cb001e51

File: src/js/config.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+// index.js
+function sayHello(name) {
+  if (name) {
+    console.log(`Hello, ${name}!`);
+  } else {
+    console.log('Hello, world!');
+  }
+}
+sayHello('Alice');
+sayHello();
```
Comment (khusharth) [inline src/js/config.js:10]: is this required?

**Comment by khusharth** [comment]
@Pranav-Patani Need to change the target branch to `main`

File: src/js/config.js (hunk 9 above)
Comment (Pranav-Patani) [reply src/js/config.js]: fixed https://github.com/Pranav-Patani/BookWise/pull/4/commits/29b8309f4516034c9aeae3e8251460c2c7a30231
It was a test code to check linting

**Review by khusharth — APPROVED** [review]
